s3/s3_operations.cpp

```diff
@@ -401,22 +402,57 @@ namespace irods_s3 {
                 }
             }
 
+            // On redirect there is not an entry in L1desc[]. The following rules explain the behavior in
+            // this instance.
+            //
+            //   1.  s3_notify_operation() gets called on the server the client is connected to.
+            //   2.  In s3_notify_operation(), this method gets called with a L1desc[] entry so that index > 0.
+            //       The L1desc[] entry is updated along with the object->physical_path() but only if
+            //       openType == CREATE. This part ensures the database gets updated with the proper physical path.
+            //   3.  On the redirected server, s3_file_create_operation() gets called which also calls
+            //       this method. In that case there is no L1desc[] entry but object->physical_path()
+            //       needs to be updated so the file is written to the correct location in S3. Do a
+            //       GenQuery to get the object_id and use this to set the object->physical_path().
+
             if (index > 0) {
+                // There is a corresponding L1desc[] entry. Look up the object_id in it. Reverse it
+                // for the key.  Write the physical_path to the L1desc[] entry as well as object->physical_path().
 
                 std::string obj_id = boost::lexical_cast<std::string>(L1desc[index].dataObjInfo->dataId);
                 std::reverse(obj_id.begin(), obj_id.end());
 
                 // make S3 key name
                 const auto s3_key_name = fmt::format("/{}/{}/{}", bucket_name, obj_id, object_name);
 
-                // update physical path
-                logger::debug("{}:{} ({}) [[{}]] updating physical_path to {}",
-                        __FILE__, __LINE__, __FUNCTION__, thread_id, s3_key_name.c_str());
-                object->physical_path(s3_key_name);
-                strncpy(L1desc[index].dataObjInfo->filePath, s3_key_name.c_str(), MAX_NAME_LEN);
-                L1desc[index].dataObjInfo->filePath[MAX_NAME_LEN - 1] = '\0';
+                // Update physical path but only on first creation otherwise the policy that was in effect
+                // at the time the object was first created rules. 
```

Comment:
"created rules"...

does this mean, "created wins"?

Reply:
I updated it to from "rules" to "wins".